src/main/scala/is/hail/sparkextras/ContextSampledRDD.scala

```diff
@@ -0,0 +1,47 @@
+package is.hail.sparkextras
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.util.random.RandomSampler
+
+import java.util.Random
+
+import scala.reflect.ClassTag
+
+class SeededPartition(
```

Comment:
So sample isn't special, any partition should have access to deterministic, seeded random data.  This has been a long-standing problem ... everywhere ... that we're on the cusp of being able to solve.  I'd ask this get put in RVDContext.  Maybe something like `RVDContext.random: math3.random.RandomGenerator` and `def RVDContext.nextSeed() = random.nextLong()`?

This will add a little complexity to ContextRDD because now we have something like mkc: (Int) => Context where Int is the partition and the constructor closes over an array of random number seeds for that task.

HailContext should have a function to generate an per-partition array of random seeds.

Reply:
After a brief chat with Patrick, I feel like we don't need an array of random seeds. Is it not sufficient to have one seed for the `RDD` and then to seed each partition's random-number-generator with `seed + partitionIndex`?

Reply:
this is what we do for balding nichols and it seems fine there.

Reply:
Do we prefer apache commons math to `scala.util.Random`?

Reply:
> After a brief chat with Patrick, I feel like we don't need an array of random seeds. Is it not sufficient to have one seed for the RDD and then to seed each partition's random-number-generator with seed + partitionIndex?

So it depends on the RNG.  When writing BN we found that the default (Java) RNG has HIGHLY correlated sequences for nearby seeds.  This caused serious problems with the BN model.  There was a SO discussing this that I'm trying to find.  So we either needed to change the default RNG or create an array of seeds that were random (looking).

Reply:
can we take seed, add partition index, and hash it to get the partition seed?

Reply:
or add hash of partition index. I dunno.

Reply:
(non-cryptographic) hash is usually some kind of xor so this I don't think this completely solves the problem since nearby partitions will still have nearby seeds.  (But hashing the partition number with the seed generated from the global RNG isn't a bad idea although it begs the question what hash.)